iio/iio_ad713x/iio_ad713x.c

```diff
@@ -0,0 +1,305 @@
+	iio_device->attributes =
+		NULL; //iio_voltage_attributes; /* no device attribute */
+	iio_device->channels = NULL;
+
+	return iio_device;
+}
+
+/**
+ * @brief iio_ad713x_transfer_dev_to_mem() - Transfer data from device into RAM.
+ * @param iio_inst - Physical instance of a iio_axi_adc device.
+ * @param bytes_count - Number of bytes to transfer.
+ * @param ch_mask - Opened channels mask.
+ * return bytes_count or negative value in case of error.
+ */
+ssize_t iio_ad713x_transfer_dev_to_mem(void *iio_inst, size_t bytes_count,
+				       uint32_t ch_mask)
+{
+	struct iio_ad713x *iio_713x_inst;
+	ssize_t ret, bytes;
+
+	if (!iio_inst)
+		return FAILURE;
+
+	iio_713x_inst = (struct iio_ad713x *)iio_inst;
+	bytes = (bytes_count * iio_713x_inst->num_channels);
+
+	ret = spi_eng_offload_load_msg(iio_713x_inst->spi_eng_desc, iio_713x_inst->msg);
+	if (ret < 0)
+		return ret;
+
+	ret = spi_eng_transfer_multiple_msgs(iio_713x_inst->spi_eng_desc, bytes);
+	if (ret < 0)
+		return ret;
+
+	if (iio_713x_inst->dcache_invalidate_range)
+		iio_713x_inst->dcache_invalidate_range(iio_713x_inst->msg->rx_buf_addr, bytes);
+
+	return bytes_count;
+}
+
+/**
+ * @brief Read chunk of data from RAM to pbuf.
+ * Call "iio_ad713x_transfer_dev_to_mem" first.
+ * This function is probably called multiple times by libtinyiiod after a
+ * "iio_ad713x_transfer_dev_to_mem" call, since we can only read "bytes_count"
+ * bytes at a time.
+ * @param iio_inst - Physical instance of a device.
+ * @param pbuf - Buffer where value is stored.
+ * @param offset - Offset to the remaining data after reading n chunks.
+ * @param bytes_count - Number of bytes to read.
+ * @param ch_mask - Opened channels mask.
+ * @return bytes_count or negative value in case of error.
+ */
+ssize_t iio_ad713x_read_dev(void *iio_inst, char *pbuf, size_t offset,
+			    size_t bytes_count, uint32_t ch_mask)
+{
+	struct iio_ad713x *iio_713x_inst;
+	uint32_t i, j = 0, current_ch = 0;
+	uint16_t *pbuf16;
+	size_t samples;
+
+	iio_713x_inst = (struct iio_ad713x *)iio_inst;
+	pbuf16 = (uint16_t*)pbuf;
+	samples = (bytes_count * iio_713x_inst->num_channels) / hweight8(
+			  ch_mask);
+	samples /= 2; /* because of uint16_t *pbuf16 = (uint16_t*)pbuf; */
+	offset = (offset * iio_713x_inst->num_channels) / hweight8(ch_mask);
+
+	for (i = 0; i < samples; i++) {
+		if (ch_mask & BIT(current_ch)) {
+			pbuf16[j] = *(uint16_t*)(iio_713x_inst->msg->rx_buf_addr + offset + i * 2);
+			j++;
+		}
+
+		if (current_ch < iio_713x_inst->num_channels - 1)
+			current_ch++;
+		else
+			current_ch = 0;
+	}
+
+	return bytes_count;
+}
+
+/**
+ * @brief Get xml corresponding to device.
+ * @param xml - Xml containing description of a device.
+ * @param iio_dev - Structure describing a device, channels and attributes.
+ * @return SUCCESS in case of success or negative value otherwise.
+ */
+ssize_t iio_ad713x_get_xml(char** xml, struct iio_device *iio_dev)
+{
+
+	*xml = calloc(1, strlen(ad713x_xml) + 1);
```

Comment:
I think this one doesn't get freed by _unregister or _delete functions

Reply:
The free is done here:
https://github.com/analogdevicesinc/libtinyiiod/blob/1246d992fa48af4d0d7b1a388cd54ba8f27e53df/tinyiiod.c#L135

Reply:
Looks like it's done but looks like there is a mixup then. I will write more in another comment on that line because my remark belongs there.